arch/arm/src/lpc17xx_40xx/lpc17_40_pwm.c

```diff
@@ -279,19 +327,135 @@ static int pwm_timer(struct lpc17_40_pwmtimer_s *priv,
                      const struct pwm_info_s *info)
 {
   irqstate_t flags;
+  uint32_t i;
+  uint32_t ret = OK;
+  uint32_t lerval = LER0_EN;
+  uint32_t pcrval = 0;
+  uint32_t mr0_freq;
 
   flags = enter_critical_section();
 
-  putreg32(info->frequency, LPC17_40_PWM1_MR0);         /* Set PWMMR0 = number of counts */
+  mr0_freq = 1.f / info->frequency * LPC17_40_PWM_CLOCK / 10;
+
+  putreg32(mr0_freq, LPC17_40_PWM1_MR0);         /* Set PWMMR0 = number of counts */
+
+#ifndef CONFIG_PWM_MULTICHAN
   putreg32(info->duty, LPC17_40_PWM1_MR1);              /* Set PWM cycle */
+#else
+  for (i = 0; i < CONFIG_PWM_NCHANNELS; i++)
+    {
+     switch (priv->channels[i].channel)
+       {
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL1
+         case 1:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
+                             + b16HALF),
+                      LPC17_40_PWM1_MR1); /* Set PWM cycle */
+             break;
+           }
+#endif
+
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL2
+         case 2:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
+                             + b16HALF),
+                      LPC17_40_PWM1_MR2); /* Set PWM cycle */
+             break;
+           }
+#endif
+
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL3
+         case 3:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
+                             + b16HALF),
+                      LPC17_40_PWM1_MR3); /* Set PWM cycle */
+             break;
+           }
+#endif
+
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL4
+         case 4:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
+                             + b16HALF),
+                      LPC17_40_PWM1_MR4); /* Set PWM cycle */
+             break;
+           }
+#endif
+
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL5
+         case 5:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
+                             + b16HALF),
+                      LPC17_40_PWM1_MR5); /* Set PWM cycle */
+             break;
+           }
+#endif
+
+#ifdef CONFIG_LPC17_40_PWM1_CHANNEL6
+         case 6:
+           {
+             putreg32(b16toi((uint64_t)info->channels[i].duty * mr0_freq
```

Comment:
Maybe better to use `ub16_t` or `ub32_t` instead of `uint64_t`?

Reply:
Thanks for the suggestion found out that `ub16mulub16` did exactly what I needed.